File: lib/carousel/carousel.dart
import "package:angular2/angular2.dart";
import "../ng2-bootstrap-config.dart";

import "package:node_shims/js.dart";
import 'dart:async';

enum Direction { UNKNOWN, NEXT, PREV }
// todo: add animate
const NAVIGATION = const {
Ng2BootstrapTheme.BS4 :
'''
<a class="left carousel-control" (click)="prev()" [hidden]="!slides.length">
  <span class="icon-prev" aria-hidden="true"></span>
  <span class="sr-only">Previous</span>
</a>
<a class="right carousel-control" (click)="next()" [hidden]="!slides.length">
  <span class="icon-next" aria-hidden="true"></span>
  <span class="sr-only">Next</span>
</a>
  ''',
  Ng2BootstrapTheme.BS3 : '''
<a class="left carousel-control" (click)="prev()" [hidden]="!slides.length">
  <span class="glyphicon glyphicon-chevron-left"></span>
</a>
<a class="right carousel-control" (click)="next()" [hidden]="!slides.length">
  <span class="glyphicon glyphicon-chevron-right"></span>
</a>
  '''
};

@Component (selector: "carousel, [carousel]",
    inputs: const [ "interval", "noTransition", "noPause", "noWrap"],
    template: '''
<div (mouseenter)="pause()" (mouseleave)="play()" class="carousel slide">
  <ol class="carousel-indicators" [hidden]="slides.length <= 1">
     <li *ng-for="#slidez of slides" [ng-class]="{active: slidez.active === true}" (click)="select(slidez)"></li>
  </ol>
  <div class="carousel-inner"><ng-content></ng-content></div>
</div>
  ''', directives: const [ CORE_DIRECTIVES, NgClass])
class Carousel implements OnDestroy {
  bool noPause = false;

  bool noWrap;

  List<Slide> slides = [];

  Timer currentInterval;

  bool isPlaying = false;

  bool destroyed = false;

  Slide currentSlide;

  num interval;

  onDestroy() {
    this.destroyed = true;
  }

//  String get interval => _interval.toString();
//
//  set interval(String value) {
//    _interval = int.parse(value);
//    restartTimer();
//  }

  select(Slide nextSlide, [ Direction direction = Direction.UNKNOWN ]) {
    var nextIndex = nextSlide.index;
    if (identical(direction, Direction.UNKNOWN)) {
      direction =
      nextIndex > getCurrentIndex() ? Direction.NEXT : Direction.PREV;
    }
    // Prevent this user-triggered transition from occurring if there is already one in progress
    if (nextSlide != null && nextSlide != currentSlide) {
      goNext(nextSlide, direction);
    }
  }

  goNext(Slide slide, Direction direction) {
    if (destroyed) {
      return;
    }
    slide.direction = direction;
    slide.active = true;
    if (currentSlide != null) {
      currentSlide.direction = direction;
      currentSlide.active = false;
    }
    currentSlide = slide;
    // every time you change slides, reset the timer
    restartTimer();
  }

  getSlideByIndex(num index) {
    var len = slides.length;
    for (var i = 0; i < len; ++i) {
      if (identical(slides [ i ].index, index)) {
        return slides [ i ];
      }
    }
  }

  int getCurrentIndex() {
    return falsey(currentSlide) ? 0 : currentSlide.index;
  }

  next() {
    var newIndex = (getCurrentIndex() + 1) % slides.length;
    if (identical(newIndex, 0) && noWrap) {
      pause();
      return null;
    }
    return select(getSlideByIndex(newIndex), Direction.NEXT);
  }

  prev() {
    var newIndex = getCurrentIndex() - 1 < 0
        ? slides.length - 1
        : getCurrentIndex() - 1;
    if (noWrap && identical(newIndex, slides.length - 1)) {
      pause();
      return null;
    }
    return select(getSlideByIndex(newIndex), Direction.PREV);
  }

  restartTimer() {
    resetTimer();
    var intervalAux = interval.toInt();
    if (intervalAux != double.NAN && intervalAux > 0) {
      currentInterval = new Timer(new Duration(milliseconds: intervalAux), () {
        var nInterval = interval;
        if (isPlaying && intervalAux != double.NAN && nInterval > 0 &&
            truthy(slides.length)) {
          next();
        } else {
          pause();
        }
      });
    }
  }

  resetTimer() {
    if (truthy(currentInterval)) {
      currentInterval.cancel();
      currentInterval = null;
    }
  }

  play() {
    if (!isPlaying) {
      isPlaying = true;
      restartTimer();
    }
  }

  pause() {
    if (!noPause) {
      isPlaying = false;
      resetTimer();
    }
  }

  addSlide(Slide slide) {
    slide.index = slides.length;
    push(slides, slide);
    if (identical(slides.length, 1) || slide.active) {
      select(slides [ slides.length - 1 ]);
      if (identical(slides.length, 1)) {
        play();
      }
    } else {
      slide.active = false;
    }
  }

  removeSlide(Slide slide) {
    splice(slides, slide.index, 1);
    if (identical(slides.length, 0)) {
      currentSlide = null;
      return;
    }
    for (var i = 0; i < slides.length; i ++) {
      slides [ i ].index = i;
    }
  }
}

@Component (selector: "slide, [slide]",
    inputs: const [ "direction", "active", "index"],
    host: const {
      "[class.active]" : "active",
      "[class.item]" : "true",
      "[class.carousel-item]" : "true"
    },
    template: '''
  <div [ng-class]="{active: active}" class="item text-center">
    <ng-content></ng-content>
  </div>
  ''', directives: const [NgClass])
class Slide implements OnInit, OnDestroy {
  Carousel carousel;

  bool active;

  Direction direction;

  num index;

  Slide(this.carousel) {}

  onInit() {
    carousel.addSlide(this);
  }

  onDestroy() {
    carousel.removeSlide(this);
  }
}

const CAROUSEL_DIRECTIVES = const [Carousel, Slide];

File: components/carousel/carousel.ts
import {
  Component, Directive,
  OnInit, OnDestroy, Input, HostBinding,
  EventEmitter, ElementRef
} from 'angular2/core';
import { NgClass, NgFor } from 'angular2/common';

import { Ng2BootstrapConfig, Ng2BootstrapTheme } from '../ng2-bootstrap-config';

export enum Direction {UNKNOWN, NEXT, PREV}
// todo: add animate

const NAVIGATION:any = {
  [Ng2BootstrapTheme.BS4]: `
    <a class="left carousel-control" (click)="prev()" [hidden]="!slides.length">
      <span class="icon-prev" aria-hidden="true"></span>
      <span class="sr-only">Previous</span>
    </a>
    <a class="right carousel-control" (click)="next()" [hidden]="!slides.length">
      <span class="icon-next" aria-hidden="true"></span>
      <span class="sr-only">Next</span>
    </a>
  `,
  [Ng2BootstrapTheme.BS3]: `
    <a class="left carousel-control" (click)="prev()" [hidden]="!slides.length">
      <span class="glyphicon glyphicon-chevron-left"></span>
    </a>
    <a class="right carousel-control" (click)="next()" [hidden]="!slides.length">
      <span class="glyphicon glyphicon-chevron-right"></span>
    </a>
  `
};

// todo:
// (ng-swipe-right)="prev()" (ng-swipe-left)="next()"
@Component({
  selector: 'carousel',
  directives: [NgClass, NgFor],
  template: `
    <div (mouseenter)="pause()" (mouseleave)="play()" class="carousel slide">
      <ol class="carousel-indicators" [hidden]="slides.length <= 1">
         <li *ngFor="#slidez of slides" [ngClass]="{active: slidez.active === true}" (click)="select(slidez)"></li>
      </ol>
      <div class="carousel-inner"><ng-content></ng-content></div>
      ${NAVIGATION[Ng2BootstrapConfig.theme]}
    </div>
  `
})
export class Carousel implements OnDestroy {
  @Input() private noWrap:boolean;
  @Input() private noPause:boolean;
  @Input() private noTransition:boolean;

  @Input() public get interval():number {
    return this._interval;
  }

  public set interval(value:number) {
    this._interval = value;
    this.restartTimer();
  }

  private slides:Array<Slide> = [];
  private currentInterval:any;
  private isPlaying:boolean;
  private destroyed:boolean = false;
  private currentSlide:Slide;
  private _interval:number;

  public ngOnDestroy() {
    this.destroyed = true;
  }

  public select(nextSlide:Slide, direction:Direction = Direction.UNKNOWN) {
    let nextIndex = nextSlide.index;
    if (direction === Direction.UNKNOWN) {
      direction = nextIndex > this.getCurrentIndex() ? Direction.NEXT : Direction.PREV;
    }

    // Prevent this user-triggered transition from occurring if there is already one in progress
    if (nextSlide && nextSlide !== this.currentSlide) {
      this.goNext(nextSlide, direction);
    }
  }

  private goNext(slide:Slide, direction:Direction) {
    if (this.destroyed) {
      return;
    }

    slide.direction = direction;
    slide.active = true;

    if (this.currentSlide) {
      this.currentSlide.direction = direction;
      this.currentSlide.active = false;
    }

    this.currentSlide = slide;

    // every time you change slides, reset the timer
    this.restartTimer();
  }

  private getSlideByIndex(index:number) {
    let len = this.slides.length;
    for (let i = 0; i < len; ++i) {
      if (this.slides[i].index === index) {
        return this.slides[i];
      }
    }
  }

  private getCurrentIndex() {
    return !this.currentSlide ? 0 : this.currentSlide.index;
  }

  private next() {
    let newIndex = (this.getCurrentIndex() + 1) % this.slides.length;

    if (newIndex === 0 && this.noWrap) {
      this.pause();
      return;
    }

    return this.select(this.getSlideByIndex(newIndex), Direction.NEXT);
  }

  private prev() {
    let newIndex = this.getCurrentIndex() - 1 < 0 ? this.slides.length - 1 : this.getCurrentIndex() - 1;

    if (this.noWrap && newIndex === this.slides.length - 1) {
      this.pause();
      return;
    }

    return this.select(this.getSlideByIndex(newIndex), Direction.PREV);
  }

  private restartTimer() {
    this.resetTimer();
    let interval = +this.interval;
    if (!isNaN(interval) && interval > 0) {
      this.currentInterval = setInterval(() => {
        let nInterval = +this.interval;
        if (this.isPlaying && !isNaN(this.interval) && nInterval > 0 && this.slides.length) {
          this.next();
        } else {
          this.pause();
        }
      }, interval);
    }
  }

  private resetTimer() {
    if (this.currentInterval) {
      clearInterval(this.currentInterval);
      this.currentInterval = null;
    }
  }

  public play() {
    if (!this.isPlaying) {
      this.isPlaying = true;
      this.restartTimer();
    }
  }

  public pause() {
    if (!this.noPause) {
      this.isPlaying = false;
      this.resetTimer();
    }
  }

  public addSlide(slide:Slide) {
    slide.index = this.slides.length;
    this.slides.push(slide);
    if (this.slides.length === 1 || slide.active) {
      this.select(this.slides[this.slides.length - 1]);
      if (this.slides.length === 1) {
        this.play();
      }
    } else {
      slide.active = false;
    }
  }

  public removeSlide(slide:Slide) {
    this.slides.splice(slide.index, 1);

    if (this.slides.length === 0) {
      this.currentSlide = null;
      return;
    }

    for (let i = 0; i < this.slides.length; i++) {
      this.slides[i].index = i;
    }
  }
}

@Component({
  selector: 'slide',
  directives: [NgClass],
  template: `
    <div [ngClass]="{active: active}" class="item text-center">
      <ng-content></ng-content>
    </div>
  `
})
export class Slide implements OnInit, OnDestroy {
  @Input() public index:number;
  @Input() public direction:Direction;

  @HostBinding('class.active')
  @Input() public active:boolean;

  @HostBinding('class.item')
  @HostBinding('class.carousel-item')
  private addClass:boolean = true;

  constructor(private carousel:Carousel) {
  }

  public ngOnInit() {
    this.carousel.addSlide(this);
  }

  public ngOnDestroy() {
    this.carousel.removeSlide(this);
  }
}

export const CAROUSEL_DIRECTIVES:Array<any> = [Carousel, Slide];
// will be deprecated
export const carousel:Array<any> = [Carousel, Slide];
